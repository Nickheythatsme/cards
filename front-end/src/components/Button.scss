
button {
    border: none;
    border-radius: 3px;
    color: white;
    padding: 1rem;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    font-size: 1rem;
    margin: 0;
    cursor: pointer;
}
