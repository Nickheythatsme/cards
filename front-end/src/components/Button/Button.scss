@import '../../App.scss';

.btn {
    display:inline-block;
    padding:0.3em 1.2em;
    margin:0 0.3em 0.3em 0;
    border-radius:2em;
    box-sizing: border-box;
    text-decoration:none;
    font-family:'Roboto',sans-serif;
    font-weight:500;
    text-align:center;
    transition: all 200ms;
    font-size: 1em;
    cursor: pointer;
    &:focus {
        outline-width: thin;
        outline-style: auto;
    }
    &:disabled {
        cursor: not-allowed;
    }
}

.btn-clicked {
    transform: scale(0.95);
    box-shadow: inset 2px 1px 2px 1px rgba(0, 0, 0, .2);
}

.btn-mouse-over {
    transform: scale(1.05);
    box-shadow: 2px 1px 2px 1px rgba(0, 0, 0, .2);
}

@mixin button-variant($variant-name, $color-name) {
    .btn-#{$variant-name} {
        font-weight:600;
        color: theme-color("light");
        background-color: theme-color($color-name);
        border: 2px solid;
        border-color: theme-color($color-name);
        &:hover {
            background-color: darken(theme-color($color-name), 10%)
        }
        &:disabled{
            background-color: darken(theme-color($color-name), 10%)
        }
    }

    .btn-#{$variant-name}-outline {
        border: 2px solid;
        border-color: theme-color($color-name);
        color: theme-color($color-name);
        background: transparent;
        &:disabled{
            background-color: transparentize($color: theme-color($color-name), $amount: 0.9);
        }
    }
}

@include button-variant("primary", "primary");
@include button-variant("secondary", "secondary");
@include button-variant("success", "success");
@include button-variant("info", "info");
@include button-variant("warning", "warning");
@include button-variant("danger", "danger");
