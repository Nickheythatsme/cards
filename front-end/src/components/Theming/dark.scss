@import '../../App.scss';

$theme-colors: map-remove($theme-colors, "primary");

$theme-colors: map-merge(
  (
    "primary":    red,
  ),
  $theme-colors
);
