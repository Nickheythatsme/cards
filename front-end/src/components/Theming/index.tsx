export const VariantNames = [
    'primary',
    'secondary',
    'tertiary',
    'info',
    'warning',
    'danger',
    'light',
    'dark',
]
export type VariantTypes =
    | 'primary'
    | 'secondary'
    | 'tertiary'
    | 'info'
    | 'warning'
    | 'danger'
    | 'light'
    | 'dark'
