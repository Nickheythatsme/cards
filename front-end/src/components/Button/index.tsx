import Button from './Button';

export {ReactiveButton} from './Button';
export default Button;
