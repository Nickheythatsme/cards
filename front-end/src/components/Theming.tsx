import React from 'react';
import Button from './Button';


export interface ThemeType {
    primary: string,
    secondary: string,
    info: string,
    warning: string,
    success: string,
    danger: string,
    textBackground: string,
    [themeName: string]: string
}

interface AllThemesType {
    [themeName: string]: ThemeType
}

export const themes: AllThemesType = {
    light: {
        primary: '#cc5cd6',
        secondary: '#5cd6bc',
        info: '#84ffe5',
        warning: '#ffbf80',
        success: '#4dff4d',
        danger: '#ff3333',
        textBackground: '#fff'
    },
    dark: {
        primary: '#5c5cd6',
        secondary: '#5cd65c',
        info: '#80ffe5',
        warning: '#ffbf80',
        success: '#4dff4d',
        danger: '#ff3333',
        textBackground: '#000'
    },
};

export const ThemeContext = React.createContext({
    theme: themes.dark,
    toggleTheme: () => {}
});

export default class Theming extends React.Component {
    state = {
        theme: themes.dark
    }

    constructor(props: any) {
        super(props);
        this.toggleTheme = this.toggleTheme.bind(this);
    }

    toggleTheme() {
        this.setState((state: any) => {
            if (state.theme === themes.dark) {
                state.theme = themes.light;
            } else {
                state.theme = themes.dark;
            }
            return state;
        }, () => {console.log('theme changed to: ' + (this.state.theme === themes.dark ? 'dark' : 'light' ))});
    };

    render() {
        return (
            <ThemeContext.Provider value={{theme: this.state.theme, toggleTheme: this.toggleTheme}}>
                <>
                <Button variant={"primary"} onClick={this.toggleTheme}>Toggle Theme</Button>
                {this.props.children}
                </>
            </ThemeContext.Provider>
        )
    }
}
