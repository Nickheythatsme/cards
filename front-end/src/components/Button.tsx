import React from 'react';
import CSS from 'csstype';
import './Button.scss';
import {ThemeContext, ThemeType} from './Theming';

interface PropsType {
    variant?: string,
    [propName: string]: any
}

export default function Button(props: PropsType) {

    function makeStyleFromTheme(variant: string, theme: ThemeType) {
        const customButtonStyle: CSS.Properties = {
            backgroundColor: theme[variant],
        };
        if (variant.startsWith('outline-')) {
            variant = variant.replace('outline-', '');
            customButtonStyle.backgroundColor = 'none'; 
            customButtonStyle.border = '2px solid'; 
            customButtonStyle.borderColor = theme[variant]; 
            customButtonStyle.color = theme[variant]; 
        } else {
            customButtonStyle.backgroundColor = theme[variant]
            customButtonStyle.color = theme.textBackground; 
        }
        return customButtonStyle
    }

    function renderButton(theme: ThemeType) {
        const style = makeStyleFromTheme(props.variant || 'primary', theme)
        return (
            <button style={style} {...props}>
                {props.children}
            </button>
        )

    }

    return (
        <ThemeContext.Consumer>
            {({theme}) => renderButton(theme)}
        </ThemeContext.Consumer>
    )
}
