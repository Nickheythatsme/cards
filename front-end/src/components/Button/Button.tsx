import React, { useState, MouseEvent } from 'react';
import { VariantTypes } from '../Theming';
import classNames from 'classnames';
import './Button.scss';

const REACTIVE_TRANSITION_TIME = 100; // ms

interface PropTypes {
    children?: React.ReactNode;
    variant?: VariantTypes;
    outline?: boolean;
    className?: string;
    [propName: string]: any;
}

const Button = React.forwardRef<any, PropTypes>(
    (
        { outline, variant, className, ...props }: PropTypes,
        ref: React.Ref<any>
    ) => {
        const concatClassName = classNames(
            'btn',
            `btn-${variant || 'primary'}${outline ? '-outline' : ''}`,
            className
        );
        return (
            <button ref={ref} {...props} className={concatClassName}>
                {props.children}
            </button>
        );
    }
);

export const ReactiveButton = React.forwardRef<any, PropTypes>(
    ({ ...props }: PropTypes, ref: React.Ref<any>) => {
        const [isClicked, setIsClicked] = useState(false);
        const [isMouseOver, setIsMouseOver] = useState(false);
        let clickTimeout: any = null;

        let className = classNames({
            'btn-clicked': isClicked,
            'btn-mouse-over': isMouseOver,
        });

        const handleClick = (e: MouseEvent) => {
            clearTimeout(clickTimeout);
            setIsMouseOver(false);
            setIsClicked(true);
            clickTimeout = setTimeout(() => {
                setIsClicked(false);
            }, REACTIVE_TRANSITION_TIME);
            props.onClick && props.onClick(e);
        };

        const handleMouseOver = (value: boolean, e: MouseEvent) => {
            setIsMouseOver(value);
            value && props.onMouseOver && props.onMouseOver(e);
            !value && props.onMouseOut && props.onMouseOut(e);
        };

        return (
            <Button
                {...props}
                onClick={handleClick}
                onMouseOver={(e: MouseEvent) => handleMouseOver(true, e)}
                onMouseOut={(e: MouseEvent) => handleMouseOver(false, e)}
                ref={ref}
                className={className}
                disabled={props.disabled}
            >
                {props.children}
            </Button>
        );
    }
);

export default Button;
