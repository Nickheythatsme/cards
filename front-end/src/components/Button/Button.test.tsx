import React from 'react';
import Button, {ReactiveButton} from './Button';
import * as renderer from 'react-test-renderer';
import {fireEvent, render, act} from '@testing-library/react';
import {VariantNames} from '../Theming';

test('Button style changes depending on props', () => {
    const renderedSnapshot = renderer.create(
        <>
            <Button variant="info">info</Button>
            {VariantNames.map(name => (
                <div key={name}>
                    <h3>{name}</h3>
                    <Button variant={name}>Submit</Button>
                    <Button variant={name} disabled>Submit</Button>
                    <Button variant={name} outline>Submit</Button>
                    <Button variant={name} outline disabled>Submit</Button>
                    <ReactiveButton variant={name}>Submit</ReactiveButton>
                    <ReactiveButton variant={name} disabled>Submit</ReactiveButton>
                    <ReactiveButton variant={name} outline>Submit</ReactiveButton>
                    <ReactiveButton variant={name} outline disabled>Submit</ReactiveButton>
                </div>
            ))}
        </>
    ).toJSON()
    expect(renderedSnapshot).toMatchSnapshot();
});

test('Button children are rendered', () => {
    const TestComponent = ({children, ...props}: any) => (
        <>TestComponent: {children}</>
    );
    const {getByText} = render(<ReactiveButton><TestComponent>This is a test</TestComponent></ReactiveButton>);
    expect(getByText(/TestComponent: This is a test/i)).toHaveTextContent('TestComponent: This is a test');
});

test('Reactive button changes state when hovered', async () => {
    const {getByText} = render(<ReactiveButton>Button</ReactiveButton>);
    expect(getByText(/Button/i).className).toMatchSnapshot('before');
    await act(async () => {
        fireEvent.click(getByText(/Button/i));
        await new Promise((resolve, reject) => {
            setTimeout(() => {
                try {
                    act(() => {
                        expect(getByText(/Button/i).className).toMatchSnapshot('during');
                    })
                    resolve();
                } catch(err) {
                    reject(err);
                }
            }, 50)
        })
        await new Promise((resolve, reject) => {
            setTimeout(() => {
                try {
                    act(() => {
                        expect(getByText(/Button/i).className).toMatchSnapshot('after');
                    })
                    resolve();
                } catch(err) {
                    reject(err);
                }
            }, 200)
        });
    });
})
