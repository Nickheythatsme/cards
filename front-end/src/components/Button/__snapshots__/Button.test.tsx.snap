// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button style changes depending on props 1`] = `
Array [
  <button
    className="btn btn-info"
  >
    info
  </button>,
  <div>
    <h3>
      primary
    </h3>
    <button
      className="btn btn-primary"
    >
      Submit
    </button>
    <button
      className="btn btn-primary"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-primary-outline"
    >
      Submit
    </button>
    <button
      className="btn btn-primary-outline"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-primary"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-primary"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-primary-outline"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-primary-outline"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
  </div>,
  <div>
    <h3>
      secondary
    </h3>
    <button
      className="btn btn-secondary"
    >
      Submit
    </button>
    <button
      className="btn btn-secondary"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-secondary-outline"
    >
      Submit
    </button>
    <button
      className="btn btn-secondary-outline"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-secondary"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-secondary"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-secondary-outline"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-secondary-outline"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
  </div>,
  <div>
    <h3>
      tertiary
    </h3>
    <button
      className="btn btn-tertiary"
    >
      Submit
    </button>
    <button
      className="btn btn-tertiary"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-tertiary-outline"
    >
      Submit
    </button>
    <button
      className="btn btn-tertiary-outline"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-tertiary"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-tertiary"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-tertiary-outline"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-tertiary-outline"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
  </div>,
  <div>
    <h3>
      info
    </h3>
    <button
      className="btn btn-info"
    >
      Submit
    </button>
    <button
      className="btn btn-info"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-info-outline"
    >
      Submit
    </button>
    <button
      className="btn btn-info-outline"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-info"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-info"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-info-outline"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-info-outline"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
  </div>,
  <div>
    <h3>
      warning
    </h3>
    <button
      className="btn btn-warning"
    >
      Submit
    </button>
    <button
      className="btn btn-warning"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-warning-outline"
    >
      Submit
    </button>
    <button
      className="btn btn-warning-outline"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-warning"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-warning"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-warning-outline"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-warning-outline"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
  </div>,
  <div>
    <h3>
      danger
    </h3>
    <button
      className="btn btn-danger"
    >
      Submit
    </button>
    <button
      className="btn btn-danger"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-danger-outline"
    >
      Submit
    </button>
    <button
      className="btn btn-danger-outline"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-danger"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-danger"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-danger-outline"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-danger-outline"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
  </div>,
  <div>
    <h3>
      light
    </h3>
    <button
      className="btn btn-light"
    >
      Submit
    </button>
    <button
      className="btn btn-light"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-light-outline"
    >
      Submit
    </button>
    <button
      className="btn btn-light-outline"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-light"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-light"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-light-outline"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-light-outline"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
  </div>,
  <div>
    <h3>
      dark
    </h3>
    <button
      className="btn btn-dark"
    >
      Submit
    </button>
    <button
      className="btn btn-dark"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-dark-outline"
    >
      Submit
    </button>
    <button
      className="btn btn-dark-outline"
      disabled={true}
    >
      Submit
    </button>
    <button
      className="btn btn-dark"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-dark"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-dark-outline"
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
    <button
      className="btn btn-dark-outline"
      disabled={true}
      onClick={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
    >
      Submit
    </button>
  </div>,
]
`;

exports[`Reactive button changes state when hovered: after 1`] = `"btn btn-primary"`;

exports[`Reactive button changes state when hovered: before 1`] = `"btn btn-primary"`;

exports[`Reactive button changes state when hovered: during 1`] = `"btn btn-primary btn-clicked"`;
